File: src/test/java/io/protobase/f7/formulas/info/ERRORTYPETest.java
package io.protobase.f7.formulas.info;

import com.google.common.truth.Truth;
import io.protobase.f7.errors.DivException;
import io.protobase.f7.errors.NAException;
import io.protobase.f7.errors.NameException;
import io.protobase.f7.errors.NullException;
import io.protobase.f7.errors.NumException;
import io.protobase.f7.errors.ParseException;
import io.protobase.f7.errors.RefException;
import io.protobase.f7.errors.ValueException;
import io.protobase.f7.formulas.logic.EQ;
import io.protobase.f7.models.Grid;
import io.protobase.f7.testutils.TestFormula;
import org.junit.Test;

import static com.google.common.truth.Truth.assertThat;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

public class ERRORTYPETest extends TestFormula {
  @Test
  public void test() {
    assertThat(ERRORTYPE.SELF.apply(null, new NullException())).isEqualTo(1.0);
    assertThat(ERRORTYPE.SELF.apply(null, new DivException())).isEqualTo(2.0);
    assertThat(ERRORTYPE.SELF.apply(null, new ValueException())).isEqualTo(3.0);
    assertThat(ERRORTYPE.SELF.apply(null, new RefException())).isEqualTo(4.0);
    assertThat(ERRORTYPE.SELF.apply(null, new NameException())).isEqualTo(5.0);
    assertThat(ERRORTYPE.SELF.apply(null, new NumException())).isEqualTo(6.0);
    assertThat(ERRORTYPE.SELF.apply(null, new NAException())).isEqualTo(7.0);
    assertThat(ERRORTYPE.SELF.apply(null, new ParseException())).isEqualTo(8.0);
  }

  @Test
  public void test_grid() {
    assertThat(ERRORTYPE.SELF.apply(null,
        Grid.builder()
            .add(0, 0, new NullException())
            .add(0, 1, 10)
            .build())
    ).isEqualTo(1.0);
  }

  @Test
  public void test_lookup() {
    ERRORTYPE F = new ERRORTYPE(lookup, collateralLookup);
    when(collateralLookup.apply(A1, M22_RANGE)).thenReturn(new DivException());
    assertThat(F.apply(A1, M22_RANGE)).isEqualTo(2.0);
    verify(collateralLookup).apply(A1, M22_RANGE);
  }

  @Test
  public void test_nonError() {
    assertThat(ERRORTYPE.SELF.apply(null, 10.0)).isEqualTo(new NAException());
    assertThat(ERRORTYPE.SELF.apply(null, true)).isEqualTo(new NAException());
    assertThat(ERRORTYPE.SELF.apply(null, "Not an error.")).isEqualTo(new NAException());
  }

  @Test
  public void test_errorFromArgumentsMismatch() {
    Truth.assertThat(EQ.SELF.apply(null, "Too few")).isEqualTo(new NAException());
    assertThat(EQ.SELF.apply(null, "A", "B", "Too many")).isEqualTo(new NAException());
  }
}

File: src/test/js/formulas/info/ERRORTYPETest.ts
import { assert } from "chai";
import { it, describe } from "../../testutils/TestUtils";
import { stub } from "sinon";
import { DivException } from "../../../../main/js/errors/DivException";
import { F7ExceptionName } from "../../../../main/js/errors/F7ExceptionName";
import { NAException } from "../../../../main/js/errors/NAException";
import { NameException } from "../../../../main/js/errors/NameException";
import { NullException } from "../../../../main/js/errors/NullException";
import { NumException } from "../../../../main/js/errors/NumException";
import { ParseException } from "../../../../main/js/errors/ParseException";
import { RefException } from "../../../../main/js/errors/RefException";
import { ValueException } from "../../../../main/js/errors/ValueException";
import { ERRORTYPE } from "../../../../main/js/formulas/info/ERRORTYPE";
import { Grid } from "../../../../main/js/models/common/Grid";
import {
  CollateralLookupFunction,
  Computed,
  LookupFunction,
} from "../../../../main/js/models/common/Types";
import { CommonModels } from "../../testutils/CommonModels";

describe("ERRORTYPE", function () {
  it("should return code for errors", function () {
    assert.deepEqual(ERRORTYPE.SELF.run(null, new NullException()), 1);
    assert.deepEqual(ERRORTYPE.SELF.run(null, new DivException()), 2);
    assert.deepEqual(ERRORTYPE.SELF.run(null, new ValueException()), 3);
    assert.deepEqual(ERRORTYPE.SELF.run(null, new RefException()), 4);
    assert.deepEqual(ERRORTYPE.SELF.run(null, new NameException()), 5);
    assert.deepEqual(ERRORTYPE.SELF.run(null, new NumException()), 6);
    assert.deepEqual(ERRORTYPE.SELF.run(null, new NAException()), 7);
    assert.deepEqual(ERRORTYPE.SELF.run(null, new ParseException()), 8);
  });

  it("should use lookup", function () {
    const lookup = stub();
    const collateralLookup = stub();
    const F = new ERRORTYPE(lookup as LookupFunction, collateralLookup as CollateralLookupFunction);
    collateralLookup.withArgs(CommonModels.A1, CommonModels.M22_RANGE).returns(new DivException());
    assert.deepEqual(F.run(CommonModels.A1, CommonModels.M22_RANGE), 2);
    assert.isTrue(collateralLookup.calledWith(CommonModels.A1, CommonModels.M22_RANGE));
    assert.equal(collateralLookup.callCount, 1);
    assert.isTrue(lookup.notCalled);
  });

  it("should handle grids", function () {
    assert.deepEqual(
      ERRORTYPE.SELF.run(null, Grid.from<Computed>([[new NullException()], [10]])),
      1
    );
  });

  it("should return error when argument lengths are wrong", function () {
    assert.deepEqual((ERRORTYPE.SELF.run(null) as NAException).name, F7ExceptionName.NA);
    assert.deepEqual(
      (ERRORTYPE.SELF.run(null, "Too many") as NAException).name,
      F7ExceptionName.NA
    );
  });
});
